nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/NiFiAtlasClient.java

```diff
@@ -0,0 +1,537 @@
+                return new Tuple<>(new AtlasObjectId(id.getGuid(), id.getTypeName(), uniqueAttrs), entity);
+            } catch (AtlasServiceException e) {
+                return null;
+            }
+        }).filter(Objects::nonNull).collect(Collectors.toMap(Tuple::getKey, Tuple::getValue));
+    }
+
+    public void registerNiFiFlow(NiFiFlow nifiFlow) throws AtlasServiceException {
+
+        // Create parent flow entity, so that common properties are taken over.
+        final AtlasEntity flowEntity = registerNiFiFlowEntity(nifiFlow);
+
+        // Create DataSet entities those are created by this NiFi flow.
+        final Map<String, List<AtlasEntity>> updatedDataSetEntities = registerDataSetEntities(nifiFlow);
+
+        // Create path entities.
+        final Set<AtlasObjectId> remainingPathIds = registerFlowPathEntities(nifiFlow);
+
+        // Update these attributes only if anything is created, updated or removed.
+        boolean shouldUpdateNiFiFlow = nifiFlow.isMetadataUpdated();
+        if (remainingPathIds != null) {
+            flowEntity.setAttribute(ATTR_FLOW_PATHS, remainingPathIds);
+            shouldUpdateNiFiFlow = true;
+        }
+        if (updatedDataSetEntities.containsKey(TYPE_NIFI_QUEUE)) {
+            flowEntity.setAttribute(ATTR_QUEUES, updatedDataSetEntities.get(TYPE_NIFI_QUEUE));
+            shouldUpdateNiFiFlow = true;
+        }
+        if (updatedDataSetEntities.containsKey(TYPE_NIFI_INPUT_PORT)) {
+            flowEntity.setAttribute(ATTR_INPUT_PORTS, updatedDataSetEntities.get(TYPE_NIFI_INPUT_PORT));
+            shouldUpdateNiFiFlow = true;
+        }
+        if (updatedDataSetEntities.containsKey(TYPE_NIFI_OUTPUT_PORT)) {
+            flowEntity.setAttribute(ATTR_OUTPUT_PORTS, updatedDataSetEntities.get(TYPE_NIFI_OUTPUT_PORT));
+            shouldUpdateNiFiFlow = true;
+        }
+
+        logger.debug("### NiFi Flow Audit Logs START");
+        nifiFlow.getUpdateAudit().forEach(logger::debug);
+        nifiFlow.getFlowPaths().forEach((k, v) -> {
+            logger.debug("--- NiFiFlowPath Audit Logs: {}", k);
+            v.getUpdateAudit().forEach(logger::debug);
+        });
+        logger.debug("### NiFi Flow Audit Logs END");
+
+        if (shouldUpdateNiFiFlow) {
+            // Send updated entities.
+            final List<AtlasEntity> entities = new ArrayList<>();
+            final AtlasEntity.AtlasEntitiesWithExtInfo atlasEntities = new AtlasEntity.AtlasEntitiesWithExtInfo(entities);
+            entities.add(flowEntity);
+            try {
+                final EntityMutationResponse mutationResponse = atlasClient.createEntities(atlasEntities);
+                logger.debug("mutation response={}", mutationResponse);
+            } catch (AtlasServiceException e) {
+                if (e.getStatus().getStatusCode() == 404 && e.getMessage().contains("ATLAS-404-00-00B")) {
```

Comment:
Can this reference AtlasErrorCode.INSTANCE_NOT_FOUND directly instead of hardcoding "ATLAS-404-00-00B"?

Reply:
That is a good idea. Thanks!